File: Bindings/Examples/Nim/InjectDLLObf.nim
import winim/lean, os, strutils, dynlib

type
  NTSTATUS = int32
  SIZE_T = uint64
  ULONG = uint32

# Function pointer types
type
  oznbvo_655212Proc = proc(
    processHandle: HANDLE,
    baseAddress: ptr pointer,
    regionSize: ptr SIZE_T,
    allocationType: ULONG,
    protect: ULONG,
    extendedParams: pointer,
    paramCount: ULONG
  ): NTSTATUS {.stdcall.}

  yexedj_555900Proc = proc(
    processHandle: HANDLE,
    baseAddress: pointer,
    buffer: pointer,
    bufferSize: SIZE_T,
    bytesWritten: ptr SIZE_T
  ): NTSTATUS {.stdcall.}

  jljtug_682236Proc = proc(
    threadHandle: ptr HANDLE,
    desiredAccess: ULONG,
    objectAttributes: pointer,
    processHandle: HANDLE,
    startRoutine: pointer,
    argument: pointer,
    createFlags: ULONG,
    zeroBits: SIZE_T,
    stackSize: SIZE_T,
    maxStackSize: SIZE_T,
    attributeList: pointer
  ): NTSTATUS {.stdcall.}

  ezhgwv_592746Proc = proc(handle: HANDLE): NTSTATUS {.stdcall.}

proc ntSuccess(status: NTSTATUS): bool =
  status >= 0

proc injectDLL(
  processHandle: HANDLE,
  dllPath: string,
  oznbvo_655212: oznbvo_655212Proc,
  yexedj_555900: yexedj_555900Proc,
  jljtug_682236: jljtug_682236Proc,
  ezhgwv_592746: ezhgwv_592746Proc
): bool =
  let absPath = dllPath.absolutePath
  let dllPathBytes = absPath & '\0'
  let pathSize = dllPathBytes.len
  var baseAddress: pointer = nil
  var regionSize: SIZE_T = SIZE_T(pathSize)
  var status = oznbvo_655212(
    processHandle,
    addr baseAddress,
    addr regionSize,
    0x3000, # MEM_COMMIT | MEM_RESERVE
    0x40,   # PAGE_EXECUTE_READWRITE
    nil,
    0
  )
  if not ntSuccess(status):
    echo "[!] Failed to allocate memory for DLL path. Status: 0x", status.toHex
    return false
  echo "[+] Allocated DLL path memory at: 0x", cast[uint64](baseAddress).toHex
  var bytesWritten: SIZE_T = 0
  status = yexedj_555900(
    processHandle,
    baseAddress,
    cast[pointer](dllPathBytes.cstring),
    pathSize.SIZE_T,
    addr bytesWritten
  )
  if not ntSuccess(status) or bytesWritten != pathSize.SIZE_T:
    echo "[!] Failed to write DLL path. Status: 0x", status.toHex, ", Bytes written: ", bytesWritten
    return false
  echo "[+] Successfully wrote DLL path to memory"
  # Get LoadLibraryA address
  let hKernel32 = GetModuleHandleA("kernel32.dll")
  if hKernel32 == 0:
    echo "[!] Failed to get kernel32.dll handle"
    return false
  let loadLibraryA = GetProcAddress(hKernel32, "LoadLibraryA")
  if loadLibraryA == nil:
    echo "[!] Failed to get LoadLibraryA address"
    return false
  echo "[+] LoadLibraryA address: 0x", cast[uint64](loadLibraryA).toHex
  # Build x64 shellcode
  var shellcode: seq[byte] = @[
    0x48'u8, 0x83, 0xEC, 0x28,             # sub rsp, 0x28
    0x48, 0xB9                            # mov rcx,
  ]
  shellcode.add(cast[array[8, byte]](cast[uint64](baseAddress)))
  shellcode.add(@[0x48'u8, 0xB8])         # mov rax,
  shellcode.add(cast[array[8, byte]](cast[uint64](loadLibraryA)))
  shellcode.add(@[
    0xFF'u8, 0xD0,                        # call rax
    0x48, 0x83, 0xC4, 0x28,               # add rsp, 0x28
    0xC3                                  # ret
  ])
  let shellcodeSize = shellcode.len
  var shellcodeAddr: pointer = nil
  regionSize = shellcodeSize.SIZE_T
  status = oznbvo_655212(
    processHandle,
    addr shellcodeAddr,
    addr regionSize,
    0x3000,
    0x40,
    nil,
    0
  )
  if not ntSuccess(status):
    echo "[!] Failed to allocate memory for shellcode. Status: 0x", status.toHex
    return false
  echo "[+] Allocated shellcode memory at: 0x", cast[uint64](shellcodeAddr).toHex
  bytesWritten = 0
  status = yexedj_555900(
    processHandle,
    shellcodeAddr,
    unsafeAddr shellcode[0],
    shellcodeSize.SIZE_T,
    addr bytesWritten
  )
  if not ntSuccess(status) or bytesWritten != shellcodeSize.SIZE_T:
    echo "[!] Failed to write shellcode. Status: 0x", status.toHex, ", Bytes written: ", bytesWritten
    return false
  echo "[+] Successfully wrote shellcode"
  var threadHandle: HANDLE = 0
  status = jljtug_682236(
    addr threadHandle,
    0x1FFFFF, # THREAD_ALL_ACCESS
    nil,
    processHandle,
    shellcodeAddr,
    nil,
    0, 0, 0, 0,
    nil
  )
  if not ntSuccess(status) or threadHandle == 0:
    echo "[!] Failed to create remote thread. Status: 0x", status.toHex, ", Handle: 0x", threadHandle
    return false
  echo "[+] Created remote thread: 0x", threadHandle
  discard WaitForSingleObject(threadHandle, 5000)
  discard ezhgwv_592746(threadHandle)
  echo "[+] Successfully injected ", dllPath, "!"
  return true

when isMainModule:
  if paramCount() != 2:
    echo "Usage: ", paramStr(0), " <pid> <dll_path>"
    quit(1)
  let pid = paramStr(1).parseInt.uint32
  let dllPath = paramStr(2)
  let PROCESS_ALL_ACCESS = 0x1F0FFF'u32
  let hSysCaller = loadLib("SysCaller.dll")
  if hSysCaller == nil:
    echo "[!] Failed to load SysCaller.dll"
    quit(1)
  let oznbvo_655212 = cast[oznbvo_655212Proc](hSysCaller.symAddr("oznbvo_655212"))
  let yexedj_555900 = cast[yexedj_555900Proc](hSysCaller.symAddr("yexedj_555900"))
  let jljtug_682236 = cast[jljtug_682236Proc](hSysCaller.symAddr("jljtug_682236"))
  let ezhgwv_592746 = cast[ezhgwv_592746Proc](hSysCaller.symAddr("ezhgwv_592746"))
  if oznbvo_655212 == nil or yexedj_555900 == nil or jljtug_682236 == nil or ezhgwv_592746 == nil:
    echo "[!] Failed to resolve SysCaller functions"
    quit(1)
  let hProcess = OpenProcess(DWORD(PROCESS_ALL_ACCESS), WINBOOL(0), DWORD(pid))
  if hProcess == 0:
    echo "[!] Failed to open process ", pid
    quit(1)
  discard injectDLL(hProcess, dllPath, oznbvo_655212, yexedj_555900, jljtug_682236, ezhgwv_592746)
  discard CloseHandle(hProcess)

File: Bindings/Examples/C/InjectDLLObf.c
#include <windows.h>
#include <stdio.h>
#include <stdint.h>

// Type definitions for SysCaller
typedef DWORD NTSTATUS;
typedef HANDLE (__stdcall *PFN_oznbvo_655212)(
    HANDLE,
    PVOID*,
    SIZE_T*,
    DWORD,
    DWORD,
    PVOID,
    DWORD
);
typedef DWORD (__stdcall *PFN_yexedj_555900)(
    HANDLE,
    PVOID,
    PVOID,
    SIZE_T,
    SIZE_T*
);
typedef DWORD (__stdcall *PFN_jljtug_682236)(
    PHANDLE,
    DWORD,
    PVOID,
    HANDLE,
    PVOID,
    PVOID,
    DWORD,
    SIZE_T,
    SIZE_T,
    SIZE_T,
    PVOID
);
typedef DWORD (__stdcall *PFN_ezhgwv_592746)(HANDLE);

#define NT_SUCCESS(Status) ((int32_t)(Status) >= 0)

int InjectDLL(HANDLE hProcess, const char* dll_path,
    PFN_oznbvo_655212 oznbvo_655212,
    PFN_yexedj_555900 yexedj_555900,
    PFN_jljtug_682236 jljtug_682236,
    PFN_ezhgwv_592746 ezhgwv_592746)
{
    // Get absolute path
    char abs_path[MAX_PATH];
    if (!GetFullPathNameA(dll_path, MAX_PATH, abs_path, NULL)) {
        printf("[!] Failed to get full DLL path.\n");
        return 0;
    }
    size_t path_len = strlen(abs_path) + 1;
    // Allocate memory for DLL path
    PVOID base_address = NULL;
    SIZE_T region_size = path_len;
    NTSTATUS status = oznbvo_655212(
        hProcess,
        &base_address,
        &region_size,
        MEM_COMMIT | MEM_RESERVE,
        PAGE_EXECUTE_READWRITE,
        NULL,
        0
    );
    if (!NT_SUCCESS(status)) {
        printf("[!] Failed to allocate memory for DLL path. Status: 0x%08X\n", status);
        return 0;
    }
    printf("[+] Allocated DLL path memory at: 0x%p\n", base_address);
    // Write DLL path
    SIZE_T bytes_written = 0;
    status = yexedj_555900(
        hProcess,
        base_address,
        (PVOID)abs_path,
        path_len,
        &bytes_written
    );
    if (!NT_SUCCESS(status) || bytes_written != path_len) {
        printf("[!] Failed to write DLL path. Status: 0x%08X, Bytes written: %llu\n", status, (unsigned long long)bytes_written);
        return 0;
    }
    printf("[+] Successfully wrote DLL path to memory\n");
    // Get LoadLibraryA address
    HMODULE hKernel32 = GetModuleHandleA("kernel32.dll");
    if (!hKernel32) {
        printf("[!] Failed to get kernel32.dll handle\n");
        return 0;
    }
    FARPROC loadLibraryA = GetProcAddress(hKernel32, "LoadLibraryA");
    if (!loadLibraryA) {
        printf("[!] Failed to get LoadLibraryA address\n");
        return 0;
    }
    printf("[+] LoadLibraryA address: 0x%p\n", loadLibraryA);
    // Build x64 shellcode
    uint8_t shellcode[32];
    int idx = 0;
    // sub rsp, 0x28
    shellcode[idx++] = 0x48; shellcode[idx++] = 0x83; shellcode[idx++] = 0xEC; shellcode[idx++] = 0x28;
    // mov rcx, <base_address>
    shellcode[idx++] = 0x48; shellcode[idx++] = 0xB9;
    *(uint64_t*)(shellcode + idx) = (uint64_t)base_address; idx += 8;
    // mov rax, <LoadLibraryA>
    shellcode[idx++] = 0x48; shellcode[idx++] = 0xB8;
    *(uint64_t*)(shellcode + idx) = (uint64_t)loadLibraryA; idx += 8;
    // call rax
    shellcode[idx++] = 0xFF; shellcode[idx++] = 0xD0;
    // add rsp, 0x28
    shellcode[idx++] = 0x48; shellcode[idx++] = 0x83; shellcode[idx++] = 0xC4; shellcode[idx++] = 0x28;
    // ret
    shellcode[idx++] = 0xC3;
    size_t shellcode_size = idx;
    // Allocate memory for shellcode
    PVOID shellcode_addr = NULL;
    region_size = shellcode_size;
    status = oznbvo_655212(
        hProcess,
        &shellcode_addr,
        &region_size,
        MEM_COMMIT | MEM_RESERVE,
        PAGE_EXECUTE_READWRITE,
        NULL,
        0
    );
    if (!NT_SUCCESS(status)) {
        printf("[!] Failed to allocate memory for shellcode. Status: 0x%08X\n", status);
        return 0;
    }
    printf("[+] Allocated shellcode memory at: 0x%p\n", shellcode_addr);
    // Write shellcode
    bytes_written = 0;
    status = yexedj_555900(
        hProcess,
        shellcode_addr,
        shellcode,
        shellcode_size,
        &bytes_written
    );
    if (!NT_SUCCESS(status) || bytes_written != shellcode_size) {
        printf("[!] Failed to write shellcode. Status: 0x%08X, Bytes written: %llu\n", status, (unsigned long long)bytes_written);
        return 0;
    }
    printf("[+] Successfully wrote shellcode\n");
    // Create remote thread
    HANDLE hThread = NULL;
    status = jljtug_682236(
        &hThread,
        0x1FFFFF, // THREAD_ALL_ACCESS
        NULL,
        hProcess,
        shellcode_addr,
        NULL,
        0, 0, 0, 0,
        NULL
    );
    if (!NT_SUCCESS(status) || !hThread) {
        printf("[!] Failed to create remote thread. Status: 0x%08X, Handle: 0x%p\n", status, hThread);
        return 0;
    }
    printf("[+] Created remote thread: 0x%p\n", hThread);
    // Wait for thread and close handle
    WaitForSingleObject(hThread, 5000);
    ezhgwv_592746(hThread);
    printf("[+] Successfully injected %s!\n", dll_path);
    return 1;
}

int main(int argc, char* argv[])
{
    if (argc != 3) {
        printf("Usage: %s <pid> <dll_path>\n", argv[0]);
        return 1;
    }
    DWORD pid = (DWORD)atoi(argv[1]);
    const char* dll_path = argv[2];
    HMODULE hSysCaller = LoadLibraryA("SysCaller.dll");
    if (!hSysCaller) {
        printf("[!] Failed to load SysCaller.dll\n");
        return 1;
    }
    PFN_oznbvo_655212 oznbvo_655212 = (PFN_oznbvo_655212)GetProcAddress(hSysCaller, "oznbvo_655212");
    PFN_yexedj_555900 yexedj_555900 = (PFN_yexedj_555900)GetProcAddress(hSysCaller, "yexedj_555900");
    PFN_jljtug_682236 jljtug_682236 = (PFN_jljtug_682236)GetProcAddress(hSysCaller, "jljtug_682236");
    PFN_ezhgwv_592746 ezhgwv_592746 = (PFN_ezhgwv_592746)GetProcAddress(hSysCaller, "ezhgwv_592746");
    if (!oznbvo_655212 || !yexedj_555900 || !jljtug_682236 || !ezhgwv_592746) {
        printf("[!] Failed to resolve SysCaller functions\n");
        return 1;
    }
    HANDLE hProcess = OpenProcess(PROCESS_ALL_ACCESS, FALSE, pid);
    if (!hProcess) {
        printf("[!] Failed to open process %lu\n", pid);
        return 1;
    }
    InjectDLL(hProcess, dll_path, oznbvo_655212, yexedj_555900, jljtug_682236, ezhgwv_592746);
    CloseHandle(hProcess);
    return 0;
} 